File: src/main/java/com/example/demo/member/application/service/MemberService.java
package com.example.demo.member.application.service;

import com.example.demo.member.adapter.in.web.request.MemberSearchRequest;
import com.example.demo.member.application.port.in.*;
import com.example.demo.member.application.port.out.MemberCommandPort;
import com.example.demo.member.application.port.out.MemberQueryPort;
import com.example.demo.member.domain.Member;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class MemberService implements GetMembersUseCase, GetMemberUseCase, CreateMemberUseCase, UpdateMemberUseCase, DeleteMemberUseCase {

    private final MemberQueryPort memberQueryPort;
    private final MemberCommandPort memberCommandPort;
    private final PasswordEncoder passwordEncoder;

    @Override
    public Page<Member> getMembers(Pageable pageable, MemberSearchRequest memberSearchRequest) {
        Page<Member> resultMembers = memberQueryPort.searchMembers(pageable, memberSearchRequest);
        return resultMembers;
    }

    @Override
    public Member getMember(String id) {
        Member resultMember = memberQueryPort.findById(id);
        return resultMember;
    }


    @Override
    public Member createMember(Member member) {
        //비밀번호 BCrypt 인코딩
        String encodedPassword = passwordEncoder.encode(member.getPassword());
        member = member.toBuilder().password(encodedPassword).build();
        Member resultMember = memberCommandPort.save(member);
        return resultMember;
    }

    @Override
    public Member updateMember(Member member) {
        Member updateMember = memberCommandPort.update(member);
        return updateMember;
    }

    @Override
    public void deleteMember(String id) {
        memberCommandPort.softDeleteById(id);
    }
}

File: src/main/java/com/example/demo/member/application/service/MemberService.kt
package com.example.demo.member.application.service

import com.example.demo.member.application.port.`in`.*
import com.example.demo.member.application.port.out.MemberCommandPort
import com.example.demo.member.application.port.out.MemberQueryPort
import com.example.demo.member.domain.Member
import com.github.f4b6a3.ulid.UlidCreator
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service

@Service
class MemberService(
    private val memberQueryPort: MemberQueryPort,
    private val memberCommandPort: MemberCommandPort,
    private val passwordEncoder: PasswordEncoder
) : CreateMemberUseCase, GetMemberUseCase, DeleteMemberUseCase, UpdateMemberUseCase, GetMembersUseCase {

    override fun getMembers(pageable: Pageable): Page<Member> =
        memberQueryPort.searchMembers(pageable)

    override fun getMember(id: String): Member =
        memberQueryPort.findById(id)

    override fun createMember(command: CreateMemberCommand): Member {
        val encodedPassword = passwordEncoder.encode(command.password)
        val member = Member(
            id = UlidCreator.getUlid().toString(),
            email = command.email,
            password = encodedPassword,
            name = command.name,
            gender = command.gender,
            phoneNumber = command.phoneNumber,
            address = command.address,
            isDeleted = false
        )
        return memberCommandPort.save(member)
    }

    override fun updateMember(command: UpdateMemberCommand): Member {

        val current = memberQueryPort.findById(command.id)

        val updated = current.copy(
            email = command.email ?: current.email,
            phoneNumber = command.phoneNumber ?: current.phoneNumber,
            address = command.address ?: current.address,
        )

        return memberCommandPort.update(updated)
    }

    override fun deleteMember(id: String) {
        memberCommandPort.softDeleteById(id)
    }
}
